v2/client.go

```diff
@@ -0,0 +1,206 @@
+package jsonrpc
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+)
+
+var (
+	// DefaultClient is the client that fulfills the the Call method
+	DefaultClient = NewClient()
+)
+
+// Client is a JSONRPC client that faciliates the calling of methods on a
+// JSONRPC server.
+type Client struct {
+	HTTPClient *http.Client
+}
+
+// NewClient creates a new client that makes use of the http.DefaultClient as
+// it's underlying HTTPClient
+func NewClient() *Client {
+	client := &Client{
+		HTTPClient: http.DefaultClient,
+	}
+	return client
+}
+
+func (client *Client) do(req *http.Request, result interface{}) error {
+
+	rawresp, err := client.HTTPClient.Do(req)
+
+	if err != nil {
+		return err
+	}
+
+	body, err := ioutil.ReadAll(rawresp.Body)
+
+	if err != nil {
+		return err
+	}
+
+	var resp clientResponse
+
+	err = json.Unmarshal(body, &resp)
+
+	if err != nil {
+		return err
+	}
+
+	if resp.Error != nil {
+		return resp.Error
+	}
+
+	err = json.Unmarshal(resp.Result, result)
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (client *Client) doBatch(req *http.Request, batch *Batch) error {
+	rawresp, err := client.HTTPClient.Do(req)
+
+	if err != nil {
+		return err
+	}
+
+	body, err := ioutil.ReadAll(rawresp.Body)
```

Comment:
Since this schema: (read body) -> (unmarshal it into some struct) is repeated several times in the code doesn't it make more sense to wrap it in a function, also since this whole enterprise revolves around operating on JSON data maybe just create a new wrapper for handling that data (marshalling/unmarshalling, indenting etc) so you don't repeat yourself.